asmcomp/power/emit.mlp

```diff
@@ -393,8 +392,38 @@ let name_for_specific = function
 let function_name = ref ""
 (* Entry point for tail recursive calls *)
 let tailrec_entry_point = ref 0
-(* Label of glue code for calling the GC *)
-let call_gc_label = ref 0
+
+module IntSet = Stdlib.Set.Make(Stdlib.Int)
+module IntMap = Stdlib.Map.Make(Stdlib.Int)
+
+(* Labels of glue code for calling the GC.
+   There is one label per size allocated. *)
+let call_gc_labels : label IntMap.t ref = ref IntMap.empty
+                     (* size -> label *)
+
+(* Return the label of the call GC point for the given size *)
+
+let label_for_call_gc ?label_after_call_gc sz =
+  match IntMap.find_opt sz !call_gc_labels with
+  | Some lbl -> lbl
+  | None ->
+      let lbl = 
+        match label_after_call_gc with Some l -> l | None -> new_label() in
+      call_gc_labels := IntMap.add sz lbl !call_gc_labels;
+      lbl
```

Comment:
Is it really the correct logic to share the `label_after_call_gc` labels for all subsequent calls of the same size, and to not-place these labels if another label already exists in the table for this size? This explicit labelling choice comes from `asmcomp/spacetime_profiling` and I'm not sure what it does, but it looks the label is used to locate the address of an instrumented call in some runtime profiling structure, which may get confused by the sharing/erasing happening here?

Reply:
I don't claim I fully understand the "label_after_call_gc" business and the way Spacetime profiling relies on it.  The code above is patterned after the what the PowerPC code generator currently does, namely:

* The (unique) GC call point is labeled with the "label_after_call_gc" carried by the first Ialloc instruction emitted, if any; otherwise, a fresh label is used.

I don't know whether that's correct w.r.t. Spacetime --- would love to hear @mshinwell about this --- but it's what the current code does.  My code follows a very similar logic:

* The GC call point for size N is labeled with the "label_after_call_gc" carried by the first Ialloc(N) instruction emitted, if any; otherwise, a fresh label is used.

For this reason, I hope it's as good (or no worse) than the current code.  But that's for a Spacetime expert to confirm.

Reply:
On other backends (I only looked at amd64) `label_after_call_gc` is used in the `record_frame` function, which if I understand correctly emits call-specific metadata. Looking at the current PowerPC code more after your comment, my conclusion is that the handling of `label_after_call_gc` is probably already broken in its current state (different allocation points can use different labels, the latest to set the global ` call_gc_label` difference will win and the other labels will never get declared). I guess that no one ever tried Spacetime on PowerPC.

Reply:
I'll be happy to `assert (label_after_call_gc = None)` if it turns out that Spacetime was not intended to work on PowerPC.

Reply:
The mystery of Spacetime on PowerPC remains, but can be solved later.